toolchain/check/import_ref.cpp

```diff
@@ -2009,9 +2025,14 @@ class ImportRefResolver {
     // Note that the individual Finish steps can add new pending work, so keep
     // going until we have no more work to do.
     while (!pending_generics_.empty() || !pending_specifics_.empty()) {
-      while (!pending_generics_.empty()) {
-        FinishPendingGeneric(pending_generics_.pop_back_val());
+      // Process generics in the order that we added them because a later
+      // generic might refer to an earlier one.
+      // TODO: Import the generic eval block rather than calling
+      // RebuildGenericEvalBlock to rebuild it so that order doesn't matter.
+      for (size_t i = 0; i != pending_generics_.size(); ++i) {
+        FinishPendingGeneric(pending_generics_[i]);
       }
+      pending_generics_.clear();
       while (!pending_specifics_.empty()) {
```

Comment:
Would it be worth updating the specific loop to match? According to FinishPendingSpecific, it can add more specifics, so maybe a similar issue could arise?

(or if you prefer, maybe it's just worth it to have both loops read the same?)

Reply:
There is no corresponding issue for specifics, because importing them doesn't call into any code that assumes the SemIR is in a valid state, and the approach used for generics is less efficient (it might allocate more memory, for example).

I've extended the comment to try to explain a bit better why generics need more delicate handling than specifics here.

Reply:
Okay, although couldn't the current loop also be less efficient since it's a backwards iteration and the `pop_back_val` is a repeated resize instead of `clear()` once?

Reply:
It could be. My expectation is that the extra allocation cost would outweigh all the other factors, and the loop body contains enough work that the resize and reverse iteration won't be above the noise floor.